File: justViewer/project_JustViewer/src/main/java/com/kds3393/just/viewer/Music/MusicPlayerPanelView.java
package com.kds3393.just.viewer.Music;

import android.content.Context;
import android.media.MediaPlayer;
import android.os.Handler;
import android.os.Message;
import android.util.AttributeSet;
import android.view.GestureDetector;
import android.view.Gravity;
import android.view.MotionEvent;
import android.view.View;
import android.view.animation.AnticipateInterpolator;
import android.view.animation.OvershootInterpolator;
import android.widget.Button;
import android.widget.LinearLayout;
import android.widget.SeekBar;
import android.widget.SeekBar.OnSeekBarChangeListener;
import android.widget.TextView;
import android.widget.Toast;

import com.common.utils.Utils;
import com.kds3393.just.viewer.Animation.LayoutTranslateAnimation;
import com.kds3393.just.viewer.Config.SettingActivity;
import com.kds3393.just.viewer.Music.MusicService.OnMediaPlayerListener;
import com.kds3393.just.viewer.R;

import java.util.ArrayList;

import butterknife.BindView;
import butterknife.ButterKnife;

public class MusicPlayerPanelView extends LinearLayout {
	private static final String TAG = "MusicPlayerPanelView";
	
	private MusicService mService;
	
	private boolean mIsMusicPanelShow = false;

    @BindView(R.id.panel_open) Button panel_open;
    @BindView(R.id.panel_title) TextView panel_title;
    @BindView(R.id.panel_artist) TextView panel_artist;
    @BindView(R.id.panel_favo) Button panel_favo;
    @BindView(R.id.panel_current_time) TextView panel_current_time;
    @BindView(R.id.panel_seekbar) SeekBar panel_seekbar;
    @BindView(R.id.panel_duration_time) TextView panel_duration_time;
    @BindView(R.id.panel_sleep_timer) Button panel_sleep_timer;
    @BindView(R.id.panel_shuffle) Button panel_shuffle;
    @BindView(R.id.panel_prev) Button panel_prev;
    @BindView(R.id.panel_play) Button panel_play;
    @BindView(R.id.panel_next) Button panel_next;
    @BindView(R.id.panel_stop) Button panel_stop;
    @BindView(R.id.panel_del) Button panel_del;


    private LayoutTranslateAnimation mShowAnimation;
    private LayoutTranslateAnimation mHideAnimation;
    
    private int mMoveDistances;						//Open Animation시 이동 거리
    
	@Override
	public void onWindowFocusChanged(boolean hasWindowFocus) {
		super.onWindowFocusChanged(hasWindowFocus);
	}
	
	protected GestureDetector mOpenButtonGestureDetector;
	public MusicPlayerPanelView(Context context) {
		super(context);
		init();
	}
	
	public MusicPlayerPanelView(Context context, AttributeSet attrs) {
		super(context, attrs);
		init();
	}
	
	private void init() {
		mShowAnimation = new LayoutTranslateAnimation(getContext(),new AnticipateInterpolator());
		mHideAnimation = new LayoutTranslateAnimation(getContext(),new OvershootInterpolator());

		setOrientation(LinearLayout.VERTICAL);
		setGravity(Gravity.CENTER_HORIZONTAL);
		setBackgroundResource(R.drawable.z_music_panel);

		createView();
	}
	
	protected MusicLyricsView mMusicLyricsPanel;
	public void setLyricsView(MusicLyricsView lyrics) {
		mMusicLyricsPanel = lyrics;
	}
	
	protected MusicListView mMusicListPanel;
	public void setMusicList(MusicListView l) {
		mMusicListPanel = l;
	}
	
	private OnMediaPlayerListener mMediaListener;
	public void setMusicService(MusicService service) {
		mService = service;
		if (mMusicListPanel != null) {
			mMusicListPanel.setMusicService(service);
		}
		if (mService != null) {
			if (mService.getMediaPlayer().isPlaying() || mService.isPause()) {
                panel_seekbar.setMax(mService.getMediaPlayer().getDuration());
                panel_duration_time.setText(Utils.getStringForDate(mService.getMediaPlayer().getDuration(),"mm:ss"));
				Mp3Id3Parser data = mService.getMusicMetaData();
                panel_title.setText(data.mTitle);
                panel_artist.setText(data.mArtist);
				this.setVisibility(View.VISIBLE);
				if (mMusicLyricsPanel != null) {
					if (isShow())
						mMusicLyricsPanel.setVisibility(View.VISIBLE);
					mMusicLyricsPanel.setLyricsText(data.mLyrics);
				}
				
	    		if (mMusicListPanel != null) {
	    			if (isShow())
	    				mMusicListPanel.setVisibility(View.VISIBLE);
	    			mMusicListPanel.setMusicListItem(mService.getMusicList());
	    		}
			}
			
			mMediaListener = new OnMediaPlayerListener() {
				@Override
				public void onStartCommanded(ArrayList<String> array) {
					if (mMusicListPanel != null)
						mMusicListPanel.setMusicListItem(mService.getMusicList());
				}
				
				@Override
				public void onPrepared(MediaPlayer mediaplsyer, Mp3Id3Parser data) {
                    panel_title.setText(data.mTitle);
                    panel_artist.setText(data.mArtist);
                    panel_seekbar.setProgress(0);
                    panel_seekbar.setMax(mService.getMediaPlayer().getDuration());
                    panel_duration_time.setText(Utils.getStringForDate(mediaplsyer.getDuration(),"mm:ss"));
					setPlayButtonImage();
					if (MusicPlayerPanelView.this.getVisibility() != View.VISIBLE) {
						MusicPlayerPanelView.this.setVisibility(View.VISIBLE);
					}
					if (mMusicLyricsPanel != null) {
						if (isShow())
							mMusicLyricsPanel.setVisibility(View.VISIBLE);
						mMusicLyricsPanel.setLyricsText(data.mLyrics);
					}
					
		    		if (mMusicListPanel != null) {
		    			if (isShow())
		    				mMusicListPanel.setVisibility(View.VISIBLE);
		    			mMusicListPanel.onChangePlayMusic();
		    		}
		    			
				}

				@Override
				public void onCompletion(MediaPlayer mediaplsyer) {
					if (MusicPlayerPanelView.this.getVisibility() == View.VISIBLE) {
						MusicPlayerPanelView.this.setVisibility(View.GONE);
						if (mMusicLyricsPanel != null)
							mMusicLyricsPanel.setInvisibleHidePanel();
						if (mMusicListPanel != null)
							mMusicListPanel.setInvisibleHidePanel();
						setSwitchMusicPanel(false);
					}
				}

				@Override
				public void onPlay(MediaPlayer mediaplsyer, String playPath) {
                    panel_play.setBackgroundResource(R.drawable.h_media_pause);
				}
				
				@Override
				public void onPause(MediaPlayer mediaplsyer, String playPath) {
                    panel_play.setBackgroundResource(R.drawable.h_media_play);
				}

				@Override
				public void onDeleteMusicFile(String filePath) {
					if (mMusicListPanel != null) {
						mMusicListPanel.setMusicListItem(mService.getMusicList());
		    		}
				}

				@Override
				public void onSleepTimer(boolean isRun, String timeString) {
					if (isRun) {
                        panel_sleep_timer.setBackgroundResource(R.drawable.h_clock);
						Toast.makeText(getContext(), timeString + " 뒤에 음악을 종료합니다.",Toast.LENGTH_SHORT).show();
					} else {
                        panel_sleep_timer.setBackgroundResource(R.drawable.h_clock_off);
					}
				}
			};
			
			mService.setOnMediaPlayerListener(this,mMediaListener);
		}
		
	}
	
	public boolean isMusicPlaying() {
		if (mService != null && mService.getMediaPlayer() == null && mService.getMediaPlayer().isPlaying()) {
			return true;
		}
		return false;
	}
    public boolean isShow() {
    	return mIsMusicPanelShow;
    }
    
    public void setSwitchMusicPanel(boolean isShow) {
    	if (mIsMusicPanelShow == isShow)
    		return;
    	mIsMusicPanelShow = isShow;
    	if (isShow) {
    		mPanelHandler.sendEmptyMessage(SHOW_MUSIC_PANEL);
    		mShowAnimation.startUsingDistance(this, true, mMoveDistances, 400,null);
    		setPlayButtonImage();
    		if (mMusicLyricsPanel != null && mMusicLyricsPanel.isLyrics()) {
				mMusicLyricsPanel.setVisibility(View.VISIBLE);
			}
    		if (mMusicListPanel != null)
    			mMusicListPanel.setVisibility(View.VISIBLE);
    		if (mService != null && mService.isRunSleepTimer()) {
                panel_sleep_timer.setBackgroundResource(R.drawable.h_clock);
			} else {
                panel_sleep_timer.setBackgroundResource(R.drawable.h_clock_off);
			}
    	} else {
    		if (mMusicLyricsPanel != null) {
				mMusicLyricsPanel.setInvisibleHidePanel();
			}
    		if (mMusicListPanel != null)
    			mMusicListPanel.setInvisibleHidePanel();
    		mHideAnimation.startUsingDistance(this, true, -mMoveDistances, 400,null);
    	}
    }
    
    @Override
    protected void onLayout(boolean changed, int l, int t, int r, int b) {
    	super.onLayout(changed, l, t, r, b);
    	if (mMoveDistances == 0) {
        	mMoveDistances = t;
    	}
    }
    
	public void createView() {
		inflate(getContext(), R.layout.music_panel, this);
        ButterKnife.bind(this, this);

        panel_seekbar.setOnSeekBarChangeListener(mSeekListener);
		// line 0	// open btn
		mOpenButtonGestureDetector = new GestureDetector(getContext(), new GestureDetector.OnGestureListener() {
			@Override public boolean onDown(MotionEvent event) {return true;}
			@Override public void onLongPress(MotionEvent e) {}
			@Override public boolean onScroll(MotionEvent e1, MotionEvent e2,float distanceX, float distanceY) {return true;}
			@Override public void onShowPress(MotionEvent e) {}

			@Override
			public boolean onFling(MotionEvent e1, MotionEvent e2,
					float velocityX, float velocityY) {
				if (!mIsMusicPanelShow && velocityY < 0)
					setSwitchMusicPanel(true);
				else if (mIsMusicPanelShow && velocityY > 0)
					setSwitchMusicPanel(false);
				return true;
			}

			@Override
			public boolean onSingleTapUp(MotionEvent e) {
				setSwitchMusicPanel(!mIsMusicPanelShow);
				return true;
			}
		});
        panel_open.setOnTouchListener(new OnTouchListener(){
			@Override
			public boolean onTouch(View v, MotionEvent event) {
				mOpenButtonGestureDetector.onTouchEvent(event);
				return true;
			}
		});

        panel_favo.setOnClickListener(new OnClickListener(){
			@Override
			public void onClick(View arg0) {

			}
		});

        panel_sleep_timer.setOnClickListener(new OnClickListener(){
			@Override
			public void onClick(View arg0) {
				mService.showSleepTimerDialog(getContext());
			}
		});

        panel_shuffle.setOnClickListener(new OnClickListener(){
			@Override
			public void onClick(View arg0) {
				boolean isMix = !SettingActivity.getMusicListShuffle(getContext());
				SettingActivity.setMusicListShuffle(getContext(),isMix);
				setSuppleButtonImage();
				if (mService != null) {
					mService.setShuffle(isMix);
				}

				if (mMusicListPanel != null)
					mMusicListPanel.setMusicListItem(mService.getMusicList());
			}
		});
        panel_prev.setOnClickListener(new OnClickListener(){
			@Override
			public void onClick(View arg0) {
				mService.movePrev();
			}
		});
        panel_play.setOnClickListener(new OnClickListener(){
			@Override
			public void onClick(View arg0) {
				if (mService.getMediaPlayer().isPlaying()) {
					mService.pause();
                    panel_play.setBackgroundResource(R.drawable.h_media_play);
				} else {
					mService.play();
                    panel_play.setBackgroundResource(R.drawable.h_media_pause);
				}
			}
		});

        panel_next.setOnClickListener(new OnClickListener(){
			@Override
			public void onClick(View arg0) {
				mService.moveNext(false);
			}
		});
        panel_stop.setOnClickListener(new OnClickListener(){
			@Override
			public void onClick(View arg0) {
				mService.stop();
			}
		});
        panel_del.setOnClickListener(new OnClickListener(){
			@Override
			public void onClick(View arg0) {
				mService.deleteCurrentMp3();
			}
		});
	}
    
    private void setPlayButtonImage() {
    	if (mService != null) {
    		if (mService.getMediaPlayer().isPlaying())
                panel_play.setBackgroundResource(R.drawable.h_media_pause);
    		else
                panel_play.setBackgroundResource(R.drawable.h_media_play);
    	} else {
            panel_play.setBackgroundResource(R.drawable.h_media_play);
    	}
    }
    
    private void setSuppleButtonImage() {
    	if (SettingActivity.getMusicListShuffle(getContext())) {
            panel_shuffle.setBackgroundResource(R.drawable.h_media_shuffle_on);
    	} else {
            panel_shuffle.setBackgroundResource(R.drawable.h_media_shuffle_off);
    	}
    }
    
	private boolean mDragging = false;
    private OnSeekBarChangeListener mSeekListener = new OnSeekBarChangeListener() {
        public void onStartTrackingTouch(SeekBar bar) {
        	mDragging = true;
        }

        public void onProgressChanged(SeekBar bar, int progress, boolean fromuser) {
        	if (!fromuser) {
                return;
            }
        	if (mService.getMediaPlayer().isPlaying()) {
        		mService.seekTo(progress);
        	}
        }

        public void onStopTrackingTouch(SeekBar bar) {
        	mDragging = false;
        }
    };
    
    private static final int SHOW_MUSIC_PANEL = 0;
    private Handler mPanelHandler = new Handler() {
        @Override
        public void handleMessage(Message msg) {
            int pos = 0;
            switch (msg.what) {
                case SHOW_MUSIC_PANEL:
                	if (mIsMusicPanelShow && mService.getMediaPlayer() != null) {
            			pos = mService.getMediaPlayer().getCurrentPosition();
            			if (!mDragging) {
                            panel_seekbar.setProgress(pos);
            			}
                    	msg = obtainMessage(SHOW_MUSIC_PANEL);
                   		sendMessageDelayed(msg, 1000 - (pos % 1000));
            		}
                    panel_current_time.setText(Utils.getStringForDate(pos,"mm:ss"));
                	break;
            }
        }
    };
    
    public void setMusicPause() {
    	if (mService != null) {
    		mService.pause();
    	}
    }
    
    public void onResume() {
    	if (mService != null) {
    		setPlayButtonImage();
    		mService.setOnMediaPlayerListener(this,mMediaListener);
    	}
    	setSuppleButtonImage();
    }
    
    public void clear() {
    	if (mService != null) {
    		mService.setOnMediaPlayerListener(this,null);
    	}
    }
    
    
}

File: JustViewer2/app/src/main/java/com/kds3393/just/justviewer2/music/MusicPlayerPanelView.kt
package com.kds3393.just.justviewer2.music

import android.annotation.SuppressLint
import android.content.Context
import android.os.Handler
import android.os.Looper
import android.os.Message
import android.util.AttributeSet
import android.view.GestureDetector
import android.view.Gravity
import android.view.LayoutInflater
import android.view.MotionEvent
import android.view.animation.AnticipateInterpolator
import android.view.animation.OvershootInterpolator
import android.widget.LinearLayout
import android.widget.SeekBar
import android.widget.SeekBar.OnSeekBarChangeListener
import com.google.android.exoplayer2.ExoPlayer
import com.kds3393.just.justviewer2.R
import com.kds3393.just.justviewer2.activity.SettingActivity
import com.kds3393.just.justviewer2.databinding.MusicPanelBinding
import com.kds3393.just.justviewer2.music.player.MusicService
import com.kds3393.just.justviewer2.utils.Event
import common.lib.base.gone
import common.lib.base.isShow
import common.lib.base.show
import common.lib.utils.SharedBus

class MusicPlayerPanelView : LinearLayout {
    private var mService: MusicService? = null
    var isShow = false
        private set
    private var mShowAnimation: com.kds3393.just.justviewer2.animation.LayoutTranslateAnimation? = null
    private var mHideAnimation: com.kds3393.just.justviewer2.animation.LayoutTranslateAnimation? = null
    private var mMoveDistances = 0 //Open Animation시 이동 거리

    private var mOpenButtonGestureDetector: GestureDetector? = null

    constructor(context: Context?) : super(context) {
        init()
    }

    constructor(context: Context?, attrs: AttributeSet?) : super(context, attrs) {
        init()
    }

    private fun init() {
        mShowAnimation = com.kds3393.just.justviewer2.animation.LayoutTranslateAnimation(context, AnticipateInterpolator())
        mHideAnimation = com.kds3393.just.justviewer2.animation.LayoutTranslateAnimation(context, OvershootInterpolator())
        orientation = VERTICAL
        gravity = Gravity.CENTER_HORIZONTAL
        setBackgroundResource(R.drawable.z_music_panel)
        createView()
        SharedBus.register<Event.MusicFileRemove>(context) {
            if (mMusicListPanel != null) {
                mMusicListPanel!!.setMusicListItem(mService!!.mPlayerManager.musicList)
            }
        }
        SharedBus.register<Event.Music>(context) {
            if (it.state == Event.MUSIC_LIST_PREPARE) {  //play list 준비 완료
                if (mMusicListPanel != null && it.musicPathList != null) mMusicListPanel!!.setMusicListItem(it.musicPathList!!)
            } else if (it.state == ExoPlayer.STATE_READY) {     //다음 play할 음악 준비 완료
                binding.panelTitle.text = it.musicInfo?.mTitle
                binding.panelArtist.text = it.musicInfo?.mArtist
                binding.panelSeekbar.progress = 0
                binding.panelSeekbar.max = mService!!.getDuration().toInt()
                binding.panelDurationTime.text = common.lib.utils.Utils.getStringForDate(it.player?.getDuration()?:0L, "mm:ss")
                setPlayButtonImage()
                if (this@MusicPlayerPanelView.visibility != VISIBLE) {
                    this@MusicPlayerPanelView.visibility = VISIBLE
                }
                if (mMusicLyricsPanel != null) {
                    if (isShow) mMusicLyricsPanel!!.show()
                    mMusicLyricsPanel!!.setLyricsText(it.musicInfo?.mLyrics)
                }
                if (mMusicListPanel != null) {
                    if (isShow) mMusicListPanel!!.show()
                    mMusicListPanel!!.onChangePlayMusic()
                }
            } else if (it.state == Event.MUSIC_STATE_PLAY) {   //play
                binding.panelPlay.setBackgroundResource(R.drawable.h_media_pause)
            } else if (it.state == Event.MUSIC_STATE_PAUSE) {   //pause
                binding.panelPlay.setBackgroundResource(R.drawable.h_media_play)
            } else if (it.state == Event.MUSIC_STATE_END) { //모든 음악 play 완료
                if (this@MusicPlayerPanelView.isShow()) {
                    this@MusicPlayerPanelView.gone()
                    if (mMusicLyricsPanel != null) mMusicLyricsPanel!!.startSlideAnimation(isShow = false, isHide = true)
                    if (mMusicListPanel != null) mMusicListPanel!!.startSlideAnimation(isShow = false, isHide = true)
                    setSwitchMusicPanel(false)
                }
            }
        }
    }

    private var mMusicLyricsPanel: MusicLyricsView? = null
    fun setLyricsView(lyrics: MusicLyricsView?) {
        mMusicLyricsPanel = lyrics
    }

    private var mMusicListPanel: MusicListView? = null
    fun setMusicList(l: MusicListView?) {
        mMusicListPanel = l
    }

    lateinit var binding: MusicPanelBinding
    fun setMusicService(service: MusicService?) {
        mService = service
        if (mMusicListPanel != null) {
            mMusicListPanel!!.setMusicService(service)
        }
        mService?.apply {
            if (mPlayerManager.isPlaying() || isPause) {
                binding.panelSeekbar.max = mPlayerManager.getDuration().toInt()
                binding.panelDurationTime.text = common.lib.utils.Utils.getStringForDate(mPlayerManager.getDuration(), "mm:ss")
                val data = mPlayerManager.musicMetaData
                binding.panelTitle.text = data?.mTitle
                binding.panelArtist.text = data?.mArtist
                this@MusicPlayerPanelView.visibility = VISIBLE
                if (mMusicLyricsPanel != null) {
                    if (isShow) mMusicLyricsPanel?.show()
                    mMusicLyricsPanel!!.setLyricsText(data?.mLyrics)
                }
                if (mMusicListPanel != null) {
                    if (isShow) mMusicListPanel?.show()
                    mMusicListPanel!!.setMusicListItem(mService!!.mPlayerManager.musicList)
                }
            }
        }
    }

    fun setSwitchMusicPanel(isShow: Boolean) {
        if (this.isShow == isShow) return
        this.isShow = isShow
        if (isShow) {
            mPanelHandler.sendEmptyMessage(SHOW_MUSIC_PANEL)
            mShowAnimation!!.startUsingDistance(this, true, mMoveDistances, 400, null)
            setPlayButtonImage()
            if (mMusicLyricsPanel != null && mMusicLyricsPanel!!.isLyrics) {
                mMusicLyricsPanel!!.visibility = VISIBLE
            }
            if (mMusicListPanel != null) mMusicListPanel!!.visibility = VISIBLE
        } else {
            if (mMusicLyricsPanel != null) {
                mMusicLyricsPanel!!.startSlideAnimation(isShow = false, isHide = true)
            }
            if (mMusicListPanel != null) mMusicListPanel!!.startSlideAnimation(isShow = false, isHide = true)
            mHideAnimation!!.startUsingDistance(this, true, -mMoveDistances, 400, null)
        }
    }

    override fun onLayout(changed: Boolean, l: Int, t: Int, r: Int, b: Int) {
        super.onLayout(changed, l, t, r, b)
        if (mMoveDistances == 0) {
            mMoveDistances = t
        }
    }

    @SuppressLint("ClickableViewAccessibility")
    private fun createView() {
        binding = MusicPanelBinding.inflate(LayoutInflater.from(context), this)
        binding.panelSeekbar.setOnSeekBarChangeListener(mSeekListener)        // line 0	// open btn
        mOpenButtonGestureDetector = GestureDetector(context, object : GestureDetector.OnGestureListener {
            override fun onDown(event: MotionEvent): Boolean {
                return true
            }

            override fun onLongPress(e: MotionEvent) {}
            override fun onShowPress(e: MotionEvent) {}
            override fun onScroll(e1: MotionEvent?, e2: MotionEvent, distanceX: Float, distanceY: Float): Boolean {
                return true
            }
            override fun onFling(e1: MotionEvent?, e2: MotionEvent, velocityX: Float, velocityY: Float): Boolean {
                if (!isShow && velocityY < 0) setSwitchMusicPanel(true) else if (isShow && velocityY > 0) setSwitchMusicPanel(false)
                return true
            }

            override fun onSingleTapUp(e: MotionEvent): Boolean {
                setSwitchMusicPanel(!isShow)
                return true
            }
        })
        binding.panelOpen.setOnTouchListener { _, event ->
            mOpenButtonGestureDetector!!.onTouchEvent(event)
            true
        }
        binding.panelFavo.setOnClickListener { }
        binding.panelShuffle.setOnClickListener {
            val isMix = !SettingActivity.getMusicListShuffle(context)
            SettingActivity.setMusicListShuffle(context, isMix)
            setSuppleButtonImage()
            if (mService != null) {
                mService!!.setShuffle(isMix)
            }
            if (mMusicListPanel != null) mMusicListPanel!!.setMusicListItem(mService!!.mPlayerManager.musicList)
        }
        binding.panelPrev.setOnClickListener { mService!!.movePrev() }
        binding.panelPlay.setOnClickListener {
            if (mService!!.isPlaying) {
                mService!!.pause()
                binding.panelPlay.setBackgroundResource(R.drawable.h_media_play)
            } else {
                mService!!.play()
                binding.panelPlay.setBackgroundResource(R.drawable.h_media_pause)
            }
        }
        binding.panelNext.setOnClickListener { mService!!.moveNext(false) }
        binding.panelStop.setOnClickListener { mService!!.stop() }
        binding.panelDel.setOnClickListener { mService!!.deleteCurrentMp3() }
    }

    private fun setPlayButtonImage() {
        if (mService != null) {
            if (mService!!.isPlaying) binding.panelPlay.setBackgroundResource(R.drawable.h_media_pause) else binding.panelPlay.setBackgroundResource(R.drawable.h_media_play)
        } else {
            binding.panelPlay.setBackgroundResource(R.drawable.h_media_play)
        }
    }

    private fun setSuppleButtonImage() {
        if (SettingActivity.getMusicListShuffle(context)) {
            binding.panelShuffle.setBackgroundResource(R.drawable.h_media_shuffle_on)
        } else {
            binding.panelShuffle.setBackgroundResource(R.drawable.h_media_shuffle_off)
        }
    }

    private var mDragging = false
    private val mSeekListener: OnSeekBarChangeListener = object : OnSeekBarChangeListener {
        override fun onStartTrackingTouch(bar: SeekBar) {
            mDragging = true
        }

        override fun onProgressChanged(bar: SeekBar, progress: Int, fromuser: Boolean) {
            if (!fromuser) {
                return
            }
            if (mService!!.isPlaying) {
                mService!!.seekTo(progress.toLong())
            }
        }

        override fun onStopTrackingTouch(bar: SeekBar) {
            mDragging = false
        }
    }
    private val mPanelHandler: Handler = object : Handler(Looper.getMainLooper()) {
        override fun handleMessage(handleMsg: Message) {
            var msg = handleMsg
            val pos : Long
            when (msg.what) {
                SHOW_MUSIC_PANEL -> {
                    pos = mService?.getCurrentPosition()?:-1
                    if (pos >= 0) {
                        if (isShow) {
                            if (!mDragging) {
                                binding.panelSeekbar.progress = pos.toInt()
                            }
                            msg = obtainMessage(SHOW_MUSIC_PANEL)
                            sendMessageDelayed(msg, (1000 - pos % 1000))
                        }
                        binding.panelCurrentTime.text = common.lib.utils.Utils.getStringForDate(pos, "mm:ss")
                    } else {
                        sendMessageDelayed(msg, 1000)
                    }
                }
            }
        }
    }

    fun onResume() {
        if (mService != null) {
            setPlayButtonImage()
        }
        setSuppleButtonImage()
    }

    fun clear() {
    }

    companion object {
        private const val SHOW_MUSIC_PANEL = 0
    }
}